src/dune/super_context.ml

```diff
@@ -654,16 +679,31 @@ module Deps = struct
       Expander.with_record_no_ddeps expander forms ~dep_kind:Optional
         ~map_exe:Fn.id ~foreign_flags
     in
-    let+ deps =
-      Build.map ~f:List.concat (List.map l ~f:(f t expander) |> Build.all)
-    and+ () = Build.record_lib_deps (Expander.Resolved_forms.lib_deps forms)
-    and+ () = Build.path_set (Expander.Resolved_forms.sdeps forms) in
-    if Pform.Expansion.Map.is_empty (Expander.Resolved_forms.ddeps forms) then
-      deps
-    else
+    match
+      Pform.Expansion.Map.is_empty (Expander.Resolved_forms.ddeps forms)
+    with
+    | false ->
       (* calling [with_record_no_ddeps] above guarantees this never happens *)
       Code_error.raise "ddeps are not allowed in this position"
         [ ("forms", Dyn.Encoder.opaque forms) ]
+    | true ->
+      let deps =
+        match all_expanded l ~f:(f t expander) with
+        | Expanded deps -> Build.all deps
+        | Unexpanded sw ->
+          Build.fail
+            { fail =
+                (fun () ->
+                  let loc = String_with_vars.loc sw in
+                  User_error.raise ~loc
+                    [ Pp.text "failed to expand percent form" ])
```

Comment:
The user should only see this error if they used some unknown percent form. Because the errors discovered before the expansion are prefixed.